// coc.nvim/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executable = exports.isRunning = exports.runCommand = exports.wait = exports.concurrent = exports.disposeAll = exports.BasicList = exports.listManager = exports.extensions = exports.FileSystemWatcher = exports.Document = exports.diagnosticManager = exports.languages = exports.sources = exports.commands = exports.services = exports.events = exports.snippetManager = exports.workspace = exports.ansiparse = exports.FloatBuffer = exports.download = exports.fetch = exports.FloatFactory = exports.Emitter = exports.Event = exports.Disposable = exports.Uri = exports.Watchman = exports.Mru = exports.Highligher = exports.Window = exports.Buffer = exports.Neovim = void 0;
const tslib_1 = require("tslib");
const commands_1 = tslib_1.__importDefault(require("./commands"));
exports.commands = commands_1.default;
const events_1 = tslib_1.__importDefault(require("./events"));
exports.events = events_1.default;
const languages_1 = tslib_1.__importDefault(require("./languages"));
exports.languages = languages_1.default;
const document_1 = tslib_1.__importDefault(require("./model/document"));
exports.Document = document_1.default;
const mru_1 = tslib_1.__importDefault(require("./model/mru"));
exports.Mru = mru_1.default;
const floatBuffer_1 = tslib_1.__importDefault(require("./model/floatBuffer"));
exports.FloatBuffer = floatBuffer_1.default;
const floatFactory_1 = tslib_1.__importDefault(require("./model/floatFactory"));
exports.FloatFactory = floatFactory_1.default;
const fetch_1 = tslib_1.__importDefault(require("./model/fetch"));
exports.fetch = fetch_1.default;
const download_1 = tslib_1.__importDefault(require("./model/download"));
exports.download = download_1.default;
const highligher_1 = tslib_1.__importDefault(require("./model/highligher"));
exports.Highligher = highligher_1.default;
const fileSystemWatcher_1 = tslib_1.__importDefault(require("./model/fileSystemWatcher"));
exports.FileSystemWatcher = fileSystemWatcher_1.default;
const services_1 = tslib_1.__importDefault(require("./services"));
exports.services = services_1.default;
const sources_1 = tslib_1.__importDefault(require("./sources"));
exports.sources = sources_1.default;
const workspace_1 = tslib_1.__importDefault(require("./workspace"));
exports.workspace = workspace_1.default;
const extensions_1 = tslib_1.__importDefault(require("./extensions"));
exports.extensions = extensions_1.default;
const manager_1 = tslib_1.__importDefault(require("./list/manager"));
exports.listManager = manager_1.default;
const manager_2 = tslib_1.__importDefault(require("./snippets/manager"));
exports.snippetManager = manager_2.default;
const basic_1 = tslib_1.__importDefault(require("./list/basic"));
exports.BasicList = basic_1.default;
const manager_3 = tslib_1.__importDefault(require("./diagnostic/manager"));
exports.diagnosticManager = manager_3.default;
const ansiparse_1 = require("./util/ansiparse");
Object.defineProperty(exports, "ansiparse", { enumerable: true, get: function () { return ansiparse_1.ansiparse; } });
const watchman_1 = tslib_1.__importDefault(require("./watchman"));
exports.Watchman = watchman_1.default;
const vscode_uri_1 = require("vscode-uri");
Object.defineProperty(exports, "Uri", { enumerable: true, get: function () { return vscode_uri_1.URI; } });
const neovim_1 = require("@chemzqm/neovim");
Object.defineProperty(exports, "Neovim", { enumerable: true, get: function () { return neovim_1.Neovim; } });
Object.defineProperty(exports, "Buffer", { enumerable: true, get: function () { return neovim_1.Buffer; } });
Object.defineProperty(exports, "Window", { enumerable: true, get: function () { return neovim_1.Window; } });
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
Object.defineProperty(exports, "Disposable", { enumerable: true, get: function () { return vscode_languageserver_protocol_1.Disposable; } });
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return vscode_languageserver_protocol_1.Event; } });
Object.defineProperty(exports, "Emitter", { enumerable: true, get: function () { return vscode_languageserver_protocol_1.Emitter; } });
tslib_1.__exportStar(require("./types"), exports);
tslib_1.__exportStar(require("./language-client"), exports);
tslib_1.__exportStar(require("./provider"), exports);
var util_1 = require("./util");
Object.defineProperty(exports, "disposeAll", { enumerable: true, get: function () { return util_1.disposeAll; } });
Object.defineProperty(exports, "concurrent", { enumerable: true, get: function () { return util_1.concurrent; } });
Object.defineProperty(exports, "wait", { enumerable: true, get: function () { return util_1.wait; } });
Object.defineProperty(exports, "runCommand", { enumerable: true, get: function () { return util_1.runCommand; } });
Object.defineProperty(exports, "isRunning", { enumerable: true, get: function () { return util_1.isRunning; } });
Object.defineProperty(exports, "executable", { enumerable: true, get: function () { return util_1.executable; } });
